File: goWork/edabit_medium_challenges.go
package goWork

import "regexp"

// Create a function that accepts a list of booleans and returns a count for how many are true
func CountTrue(values []bool) int {
	var counter = 0
	for i := 0; i < len(values); i++ {
		if values[i] {
			counter += 1
		}
	}
	return counter
}

// Create a function that accepts a string and returns a function that returns that string
func Redundant(value string) func() string {
	return func() string { return value }
}

// Create a RegEx to check for an empty string
// Compile regex outside of method call since regex compilation is expensive
var emptyRegEx, _ = regexp.Compile("^$")
func RegExpEmptyString(value string) bool {
	return emptyRegEx.MatchString(value)
}

// Create a function that takes two numbers as arguments (num, length) and
// returns an slice of multiples of num until the list length reaches length.
func SliceOfMultiples(num int, length int) []int {
	var multiple = []int{}
	for i := 1; i <= length; i++ {
		multiple = append(multiple, num*i)
	}
	return multiple
}

// Given a list of positive numbers, sum up the Odds and Even numbers separately and
// return the difference (larger sum - smaller sum)
func WarOfNumbers(numbers []int) int {
	var evenSum = 0
	var oddSum = 0
	for i := 0; i < len(numbers); i++ {
		if numbers[i]%2 == 0 {
			evenSum += numbers[i]
		} else {
			oddSum += numbers[i]
		}
	}
	return Abs(evenSum - oddSum)
}

// Utility function for War Of Numbers
func Abs(num int) int {
	if num < 0 {
		return -num
	}
	return num
}

// Create a function that finds the sum of an Slice recursively
func SquarePatch(value int) [][]int {
	var internalList = []int{}
	var externalList = [][]int{}

	// Create internal slice
	for i := 0; i < value; i++ {
		internalList = append(internalList, value)
	}

	// Build external slice
	for i := 0; i < value; i++ {
		externalList = append(externalList, internalList)
	}

	return externalList
}

// Create a function that finds the sum of an slice recursively
func RecursiveSum(values []int) int {
	switch len(values) {
	case 0:
		return 0
	case 1:
		return values[0]
	default:
		return values[0] + RecursiveSum(values[1:])
	}
}

File: scalaWork/src/main/scala/scalaSolutions/EdabitMediumChallenges.scala
package scalaSolutions

import scala.util.matching.Regex

object EdabitMediumChallenges {

  // Create a function that accepts a list of booleans and returns a count for how many are true
  def countTrue(booleanList: List[Boolean]): Int = {
    booleanList.map{bool: Boolean =>
    if (bool)
      1
    else
      0
    }.sum
  }

  // Create a function that accepts a string and returns a function that returns that string
  def redundant(value: String): ()=>String = {
    () => value
  }

  // Create a RegEx to check for an empty string
  // Compile regex outside of method call since regex compilation is expensive
  val emptyRegEx: Regex = """^$""".r // bad scope
  def regExEmptyString(value: String): Boolean = emptyRegEx.matches(value)

  // Create a function that takes two numbers as arguments (num, length) and
  // returns an array of multiples of num until the list length reaches length.
  def arrayOfMultiples(num: Int, len: Int): List[Int] = {
    (1 to len).map {value =>
      num * value
    }.toList
  }

  // Given a list of positive numbers, sum up the Odds and Even numbers separately and
  // return the difference (larger sum - smaller sum)
  def warOfNumbers(numbers: List[Int]): Int = {
    // Sum even numbers
    val evenSum = numbers.map{ num =>
      if (num % 2 == 0)
        num
      else
        0
    }.sum

    // Sum odd numbers
    val oldSum = numbers.map { num =>
      if (num % 2 != 0)
        num
      else
        0
    }.sum

    math.abs(oldSum - evenSum)
  }

  // Create a function that takes a 0 or positive integer and outputs an n x n squar consisting of only integer n
  def squarePatch(value: Int):List[List[Int]] = {
    val internalList: List[Int] = (1 to value).map{num: Int => value}.toList
    (1 to value).map{num:Int => internalList}.toList
  }

  // Create a function that finds the sum of an List recursively
  def recursiveSum(values: List[Int]): Int = {
    values.length match {
      case 0 => 0
      case 1 => values.head
      case _ => values.head + recursiveSum(values.tail)
    }
  }

  // Create a function that counts the number of times a character appears in a List[List[Char]]
  def letterCount(charLists: List[List[Char]], searchChar: Char): Int = {
    charLists.flatten.map{char: Char =>
      if (char == searchChar)
        1
      else
        0
    }.sum
  }
  
}
